ringbuffer.py

```diff
@@ -185,23 +189,51 @@ def try_write(self, data):
             self.array[position.index] = data
             self.writer.increment()
 
-    def _has_read_conflict(self, reader_position):
+    def _try_read_no_lock(self, reader: Pointer, length: int = 1):
+        position = reader.get()
         writer_position = self.writer.get()
-        return writer_position.counter <= reader_position.counter
 
-    def _try_read_no_lock(self, reader):
-        position = reader.get()
-        if self._has_read_conflict(position):
+        if writer_position.counter <= position.counter:
             if not self.active.value:
                 raise WriterFinishedError
             else:
                 raise WaitingForWriterError
 
-        data = self.array[position.index]
-        reader.increment()
-        return data
-
-    def try_read(self, reader):
+        max_length = writer_position.counter - position.counter
+
+        if max_length < length:
+            raise InsufficientDataError
+
+        remaining_buffer_length = self.slot_count - position.index
+        new_array = (self.c_type * length)()
+        if position.generation == writer_position.generation or \
+                remaining_buffer_length > length:
+            # If the writer is ahead of the reader in the current buffer
+            # generation, a simple memmove can copy the array.
+            ctypes.memmove(new_array,
+                           ctypes.addressof(self.array[position.index]),
+                           ctypes.sizeof(new_array))
+        else:
+            # When the writer has a greater generation count, two memmove
+            # calls are necessary to fill the new array with data.
+            first_slice_length = remaining_buffer_length
+            second_slice_length = length - first_slice_length
+            one_item_size = int(ctypes.sizeof(new_array) / length)
```

Comment:
Is this different from `ctypes.sizeof(self.c_type)`?

Reply:
good catch, simplifying this.